hive/src/main/java/org/apache/iceberg/hive/ClientPool.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.hive;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Deque;
+import java.util.function.Consumer;
+
+abstract class ClientPool<C, E extends Exception> implements Closeable {
```

Comment:
Not yet. I thought it may be useful in the future and shouldn't be tied to Hive, but I'm not sure there will be other uses so I kept it package-private for now.

Reply:
I don't think I agree about using `AutoCloseable`. Two reasons:
1. Close will close client connections, which do release IO streams. So this is an IO stream close.
2. Even if `AutoCloseable` were a better fit for exceptions, `Closeable` is the one more often used. I even wrote the initial version of close by checking whether a client implemented `Closeable` to call close. That wouldn't work with `AutoCloseable`. I don't think people actually think to use `AutoCloseable` in those cases, so implementing `Closeable` is more likely to work.

Reply:
1. The difference between `Closeable` and `AutoCloseable` is only in the type of exception thrown by the `close()` method. In the case of `ClientPool`, `close()` does not throw `IOException`, so it is closer to `AutoCloseable`.
1. `ClientPool` does not close IO streams by itself. It releases resources (connections).
1. It is arguable what is more often used as every class that implements `Closeable` also implements `AutoCloseable`.
1. The question is what interface `ClientPool` implements. An individual resource/client/connection that `ClientPool` manages may or may not implement `Closeable` or `AutoCloseable` at all. For example in case of `HiveMetaStoreClient` it implements `IMetaStoreClient` that is neither of those.

Reply:
You're right that this doesn't currently wrap a `Closeable`, but I think that because `Closeable` is the generally expected interface, it is better to implement that one. `AutoCloseable` didn't exist until Java 7 and `Closeable` extends `AutoCloseable`, so the one that typically gets used is `Closeable`. It may well be more correct to use `AutoCloseable`, but it's a minor distinction and this works in more cases.

Reply:
I agree that it's a minor distinction (so was the suggestion to consider switching to `AutoCloseable`, not a request for the code change), but I don't agree that `Closeable` works in *more* cases.

Reply:
Thanks for the context. By more cases, I mean this: all instances of `Closeable` are `AutoCloseable` and people are used to handling `Closeable` and may not check for `AutoCloseable`. So there are cases where callers expect `Closeable` and returning an `AutoCloseable` won't work. I think the logic holds because all `Closeable` instances are `AutoCloseable`, but not vice versa. Anyway, it's minor and this isn't even public so it doesn't matter.

Reply:
I don't agree with that logic. It is incorrect to expect that everything is `Closeable`. According to java doc, *A `Closeable` is a source or destination of data that can be closed.* `ClientPool` is neither a source or a destination of data. It is a managed collection of clients. IMO, it is incorrect for an object to pretend that it is a `cat` when it is an `animal` due to other objects being only familiar how to deal with cats :).

Yes, it's a minor comment, not a blocker for the PR to be merged.

Reply:
An important distinction is that `Closeable` is required to be idempotent but `AutoCloseable` is not.